### PR TITLE
Initial port of Node.js buffer tests and fixes

### PR DESCRIPTION
**Pull request by jasnell**
Note that this pulls in parts of the implementation directly from Node.js. Namely, the StringSearch and base64 decoding impls. See comments in the C++ files for details.

### CODE REVIEW

File: src/workerd/api/node/buffer.c++ [hunk 1]
```diff
@@ -105,12 +127,15 @@ kj::Maybe<uint> tryFromHexDigit(char c) {
...
 
-kj::Array<byte> decodeHexTruncated(kj::ArrayPtr<const char> text) {
+kj::Array<byte> decodeHexTruncated(kj::ArrayPtr<kj::byte> text, bool strict = false) {
   // We do not use kj::decodeHex because we need to match Node.js'
   // behavior of truncating the response at the first invalid hex
   // pair as opposed to just marking that an error happened and
   // trying to continue with the decode.
   if (text.size() % 2 != 0) {
+    if (strict) {
+      JSG_FAIL_REQUIRE(TypeError, "Text text is not valid hex");
```
Comment (dom96) [inline src/workerd/api/node/buffer.c++:137]:
```suggestion
      JSG_FAIL_REQUIRE(TypeError, "The text is not valid hex");
```

File: src/workerd/api/node/buffer.c++ [hunk 2]
```diff
@@ -120,18 +145,128 @@ kj::Array<byte> decodeHexTruncated(kj::ArrayPtr<const char> text) {
...
+
+uint32_t writeInto(
+    jsg::Lock& js,
+    kj::ArrayPtr<kj::byte> buffer,
+    v8::Local<v8::String> string,
+    uint32_t offset,
+    uint32_t length,
+    Encoding encoding) {
+  auto dest = buffer.slice(offset, kj::min(offset + length, buffer.size()));
+  if (dest.size() == 0 || string->Length() == 0) return 0;
```
Comment (dom96) [inline src/workerd/api/node/buffer.c++:175]: Not sure if our style guide calls for this, but I've seen this style lead to bugs so nit:

```suggestion
  if (dest.size() == 0 || string->Length() == 0) { return 0; }
```

File: src/node/internal/internal_buffer.ts [hunk 3]
```diff
@@ -711,25 +753,38 @@ Buffer.prototype.write = function write(string: StringLike,
...
-  if (arguments.length === 1) {
+  if (offset === undefined) {
     // Buffer#write(string)
     return bufferUtil.write(this, string as string, 0, this.length, "utf8");
   }
 
-  if (typeof offset === 'string') {
-    encoding = offset as string;
-    offset = 0;
+  if (length === undefined && typeof offset === 'string') {
```
Comment (dom96) [inline src/node/internal/internal_buffer.ts:761]: I know this is how Node does things, but this kind of function overloading emulation sure makes me uneasy.

File: src/workerd/api/node/buffer-base64.h [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+#include <cmath>
+#include <cstddef>
+#include <cstdint>
+
+
+namespace workerd::api::node {
+
+// This is Node.js base64 implementation. We use this instead of kj's for Node.js buffer
+// so that decoding matches the Node.js semantics.
+// TODO(later): Reconcile this with kj so we don't have to duplicate.
```
Comment (dom96) [inline src/workerd/api/node/buffer-base64.h:36]: Hmm, in what way does Node and kj differ here? Surely base64 encoding/decoding is pretty simple.

**Review by dom96 — APPROVED** [review]
Is there a lot of changes in the source code taken from Node? Should/could we be depending on these sources from Node's git directly?

Changes LGTM, I would push back on the Node base64 inclusion but I can see you are planning to remove it later so I don't see a reason to block this PR on that.

File: src/node/internal/internal_buffer.ts (hunk 3 above)
Comment (jasnell) [reply src/node/internal/internal_buffer.ts:761]: Yep, it's not the best but we're rather stuck with it when implementing these apis

File: src/workerd/api/node/buffer-base64.h (hunk 4 above)
Comment (jasnell) [reply src/workerd/api/node/buffer-base64.h:36]: Node.js will ignore invalid base64 in the middle and will truncate any additional data after the padding. kj will not.

**Pull request merged**